pkg/controller/clustersync/clustersync_controller.go

```diff
@@ -0,0 +1,950 @@
+		needToCreateClusterSync = true
+	case err != nil:
+		logger.WithError(err).Log(controllerutils.LogLevel(err), "could not get ClusterSync")
+		return reconcile.Result{}, err
+	}
+
+	needToCreateLease := false
+	leaseKey := client.ObjectKey{
+		Namespace: cd.Namespace,
+		Name:      fmt.Sprintf("%s-sync", cd.Name),
+	}
+	lease := &hiveintv1alpha1.ClusterSyncLease{}
+	switch err := r.Get(context.Background(), leaseKey, lease); {
+	case apierrors.IsNotFound(err):
+		logger.Info("Lease for ClusterSync does not exist; will need to create")
+		needToCreateLease = true
+	case err != nil:
+		logger.WithError(err).Log(controllerutils.LogLevel(err), "could not create lease for ClusterSync")
+		return reconcile.Result{}, err
+	}
+
+	origStatus := clusterSync.Status.DeepCopy()
+
+	syncSets, err := r.getSyncSetsForClusterDeployment(cd, logger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	selectorSyncSets, err := r.getSelectorSyncSetsForClusterDeployment(cd, logger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	needToDoFullReapply := needToCreateClusterSync || r.timeUntilFullReapply(lease) <= 0
+	if needToDoFullReapply {
+		logger.Info("need to reapply all syncsets")
+	}
+
+	// Apply SyncSets
+	syncStatusesForSyncSets, syncSetsNeedRequeue := r.applySyncSets(
+		cd,
+		"SyncSet",
+		syncSets,
+		clusterSync.Status.SyncSets,
+		needToDoFullReapply,
+		resourceHelper,
+		logger,
+	)
+	clusterSync.Status.SyncSets = syncStatusesForSyncSets
+
+	// Apply SelectorSyncSets
+	syncStatusesForSelectorSyncSets, selectorSyncSetsNeedRequeue := r.applySyncSets(
+		cd,
+		"SelectorSyncSet",
+		selectorSyncSets,
+		clusterSync.Status.SelectorSyncSets,
+		needToDoFullReapply,
+		resourceHelper,
+		logger,
+	)
+	clusterSync.Status.SelectorSyncSets = syncStatusesForSelectorSyncSets
+
+	setFailedCondition(clusterSync)
+
+	// Create or Update the ClusterSync
+	if needToCreateClusterSync {
+		logger.Info("creating ClusterSync")
+		clusterSync.Namespace = cd.Namespace
+		clusterSync.Name = cd.Name
+		clusterSync.OwnerReferences = []metav1.OwnerReference{{
+			APIVersion:         hivev1.SchemeGroupVersion.String(),
+			Kind:               "ClusterDeployment",
+			Name:               cd.Name,
+			UID:                cd.UID,
+			BlockOwnerDeletion: pointer.BoolPtr(true),
+		}}
+		if err := r.Create(context.Background(), clusterSync); err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "could not create ClusterSync")
+			return reconcile.Result{}, err
+		}
+	}
+	if needToCreateClusterSync || !reflect.DeepEqual(origStatus, &clusterSync.Status) {
+		logger.Info("updating ClusterSync")
+		if err := r.Status().Update(context.Background(), clusterSync); err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "could not update ClusterSync")
+			return reconcile.Result{}, err
+		}
+	}
+
+	if needToDoFullReapply {
+		logger.Info("setting last full apply time")
+		lease.Spec.RenewTime = metav1.NowMicro()
+		if needToCreateLease {
```

Comment:
✔ `needToCreateLease && !needToDoFullReapply` is impossible due to the same logic noted [above](https://github.com/openshift/hive/pull/1093/files#diff-d64ff608a67db50bd6a15a35465607c1R312).